### PR TITLE
Add CI Test for Python Wrappers w/ PML

### PR DESCRIPTION
**Pull request by EZoni**
Here's some examples of the figures saved by the PICMI script of the new CI test.

Left: Ex (regular grid + guard cells)
Right: Exz in the PML (regular grid, empty, + PML cells + guard cells, some overlapping with regular grid)
<p align="center">
    
    
</p>

The script produces and saves similar figures for all E and B fields, except for the PML components `yy` and `zy` (identically zero). For reference, the test has 128 regular cells, 8 guard cells, and 10 PML cells, in each direction.

To-do:
- [x] Merge #2403 and rebase
- [x] Clean up Python code
- [x] ~~Try testing symmetry~~ (each field initialized with unit pulse: no simple symmetry)
- [x] Post sample output figures in PR's description
- [x] Link test in documentation as Python example
- [x] Maybe add plots of F,G in regular grid and PML

### CODE REVIEW

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+sim.add_diagnostic(field_diag)
+
+# Write input file to run with compiled version
+sim.write_input_file(file_name = 'inputs_2d')
+
+# Whether to include guard cells in data returned by Python wrappers
+include_ghosts = 1
+
+# Get fields data using Python wrappers
+def get_data():
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Since this is only called once, this function is not really necessary. My own personal preference is to use `global` as little as possible.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    # Plot title
+    suptitle = 'PML in (x,z), 4 grids 64 x 64'
+    plt.suptitle(suptitle)
+    figname = 'figure_' + name + '.png'
+    fig.savefig(figname, dpi = 100)
+
+# Initialize fields data (unit pulse) and apply smoothing
+def init_data(data):
+    data[nx//2,nz//2] = 1.
+    data[2:nx-1,:] *= 0.5
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Doing the smoothing over the full grid seems excessive. I think this code would be cleaner if you write out the impulse matrix and copy it to the array. For example

> impulse = np.array([1./16., 1./8.. 1/16.], [1./8., 1./4., 1./8.], [1./16., 1./8.. 1/16.]])
data[nx//2-1:nx//2+2, ny//2-1:ny//2+2] = impulse

Or something fancier

> impulse1d = np.array([1./4., 1./2., 1./4.])
impulse = np.outer(impulse1d, impulse1d)
data[nx//2-1:nx//2+2, ny//2-1:ny//2+2] = impulse

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    data[2:nx-1,:] += 0.5 * (data[1:nx-2,:] + data[3:nx,:])
+    data[:,2:nz-1] *= 0.5
+    data[:,2:nz-1] += 0.5 * (data[:,1:nz-2] + data[:,3:nz])
+
+# Advance simulation for one time step and get fields data
+sim.step(1)
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Instead of doing one step, you could also do

> sim.initialize_inputs()
sim.initialize_warpx()

Though that has the hazard that the `step` routine may change. (Though it would be best if the `Simulation` class had a separate initialization routine, but that would be a different PR.)

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+init_data(F)
+init_data(G)
+
+# Advance simulation until last time step
+sim.step(max_steps-1)
+
+# Plot E
+edge = np.array([-nxg, nx+nxg, -nzg, nz+nzg])
+# Ex
+data = Ex[-nxg:nx+nxg,-nzg:nz+nzg].transpose()
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Since the data slicing is the same for all of the main grid arrays and for all of the pml arrays, the code would be cleaner if the slicing would be done in the `plot_data` routine. It can be done in the the `if (pml)` code blocks. Similarly, you can create the `edge` inside the `plot_data` routine. (A minor note, I prefer to name variables like that with the same name as the function argument, i.e. using `extent` instead of `edge.)

**Comment by dpgrote** [comment]
Thanks for this PR! Having a CI test of the Python fields wrapper is a good idea. I have some comments in the input file, mostly stylistic.

A suggestion is that the PICMI input file could check the numerical values in the arrays. The plots are only really checking that the indexing of the arrays doesn't return an error. There is nothing to check the correctness of the plots. You can do a check comparable to the CI benchmarks, and compare the sum of the arrays as generated in Python to benchmark values that you would precalculate. The input file would use an assert, checking that the values are correct.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 1 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/fc47be1282a20b24ebb507e945c81df0fdff4f1b.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 4 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/872d235f512aa8a0ce9dce82468db77d462b9ded and https://github.com/ECP-WarpX/WarpX/pull/2576/commits/57e5155c5f8013882086a76a0db63cf9cc6ce429.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        ax.axvline(x = nx+nxpml, linewidth = lw, color = 'grey' , linestyle = ls)
+        ax.axhline(y = 0       , linewidth = lw, color = 'grey' , linestyle = ls)
+        ax.axhline(y = 0+nzg   , linewidth = lw, color = 'black', linestyle = ls)
+        ax.axhline(y = -nzpml  , linewidth = lw, color = 'grey' , linestyle = ls)
+        ax.axhline(y = nz      , linewidth = lw, color = 'grey' , linestyle = ls)
+        ax.axhline(y = nz-nzg  , linewidth = lw, color = 'black', linestyle = ls)
+        ax.axhline(y = nz+nzpml, linewidth = lw, color = 'grey' , linestyle = ls)
+        # Set extent and sliced data
+        extent = np.array([-nxg-nxpml, nx+nxpml+nxg, -nzg-nzpml, nz+nzpml+nzg])
+        X = data[-nxg-nxpml:nx+nxpml+nxg,-nzg-nzpml:nz+nzpml+nzg].transpose()
```
Comment (EZoni) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: @dpgrote It seems like this slicing on the PML arrays is not working (returns an empty array of size zero). For example, I pass `data = Expml[:,:,0]` to the function `plot_data` and then take this slice here, which doesn't work. However, if I take the same slice directly on `Expml` (without the intermediate variable `data`), then it's fine. Do you happen to know if passing through an intermediate variable changes the lower and upper bounds of these PML arrays somehow?

Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: For instance, if I set `data = Expml[:,:,0]`, then I get:
- `data[-18:-1,:].shape` returns `(17, 165)` (this one seems right)
- `data[-18:0,:].shape` returns `(0, 165)` (this one is of size zero)

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 3 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: I prefer the way you suggested, so I fixed this in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/c5c06f4115254f07762595c32b2d6b83a1fc2379.

**Comment by EZoni** [comment]
> A suggestion is that the PICMI input file could check the numerical values in the arrays. The plots are only really checking that the indexing of the arrays doesn't return an error. There is nothing to check the correctness of the plots. You can do a check comparable to the CI benchmarks, and compare the sum of the arrays as generated in Python to benchmark values that you would precalculate. The input file would use an assert, checking that the values are correct.

@dpgrote Thank you also for this suggestion, Dave. One more question: Would you say to precompute the benchmark values from the Python arrays themselves (so this would be a non-trivial check only after some code change)? I think I cannot use the values of our checksums because those are computed from the regular diagnostics, which are cell-centered, while the values computed from the Python arrays should be from the raw data, so I don't think they would be expected to match.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 5 above)
Comment (dpgrote) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: The way negative indices are handled in the wrapper objects is different than with numpy arrays. With the wrapper, the indexing is done the same way that it is done in the C++ code, where negative indices refer to guard cells. The indices are relative to having the index `[0,0]` refer to the bottom corner of the valid cells. When you do a slice of a wrapper object, it returns a numpy array. For numpy arrays however, a negative index is relative to the upper edge of the array. So when you use the index `[-18:0,:]`, it has length zero since for the numpy array `-18` is to the right of '0'.

**Comment by dpgrote** [comment]
I would precompute the sums from the Python arrays. The input file would compute the sums and you would run it locally to get the values and hard code asserts checking for those values.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        ax.annotate('PML', xy = (64,-5), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML', xy = (64,133), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (5,64), rotation = 'vertical', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (-13,64), rotation = 'vertical', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (125,64), rotation = 'vertical', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (143,64), rotation = 'vertical', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (64,4), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (64,-15), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (64,124), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (64,142), rotation = 'horizontal', ha = 'center', va = 'center')
```
Comment (EZoni) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Note: will write the locations of these annotations, and the ones below, in terms of `nx`, `nxg`, `nxpml`, etc.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 2 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Thank you for this suggestion, fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/b05e0e1814c69c7b85c78ec6202eaef6bb08de81.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+# Write input file to run with compiled version
+sim.write_input_file(file_name = 'inputs_2d')
+
+# Whether to include guard cells in data returned by Python wrappers
+include_ghosts = 1
+
+# Compute min and max of fields data
+def compute_minmax(data):
+    vmin = -max(abs(data.min()), abs(data.max()))
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: This could be done more succinctly.

> vmax = np.abs(data).max()
vmin = -vmax

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    # Set plot title
+    suptitle = 'PML in (x,z), 4 grids 64 x 64'
+    plt.suptitle(suptitle)
+    # Save figure
+    figname = 'figure_' + name + '.png'
+    fig.savefig(figname, dpi = 100)
+
+# Initialize fields data (unit pulse) and apply smoothing
+def init_data(data):
+    data[nx//2,nz//2] = 1.
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Thanks for updating this. Though, this line setting the central point to 1 is not needed since it will be overwritten below.

Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Right, sorry about that. Fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/66ec869749ffde6c8ed047729d77f002c1d5a21b (along with other things).

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 7 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Thank you for the suggestion, fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/66ec869749ffde6c8ed047729d77f002c1d5a21b (along with other things).

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+# Plot F and G in PML
+plot_data(Fpml[:,:,0], pml = True, title = 'Fx in PML', name = 'Fx')
+plot_data(Fpml[:,:,1], pml = True, title = 'Fy in PML', name = 'Fy')
+plot_data(Fpml[:,:,2], pml = True, title = 'Fz in PML', name = 'Fz')
+plot_data(Gpml[:,:,0], pml = True, title = 'Gx in PML', name = 'Gx')
+plot_data(Gpml[:,:,1], pml = True, title = 'Gy in PML', name = 'Gy')
+plot_data(Gpml[:,:,2], pml = True, title = 'Gz in PML', name = 'Gz')
+
+# Check values with benchmarks
```
Comment (EZoni) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]:
```suggestion
# Check values with benchmarks (precomputed from the same Python arrays)
```

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    figname = 'figure_' + name + '.png'
+    fig.savefig(figname, dpi = 100)
+
+# Initialize fields data (unit pulse) and apply smoothing
+def init_data(data):
+    impulse_1d = np.array([1./4., 1./2., 1./4.])
+    impulse = np.outer(impulse_1d, impulse_1d)
+    data[nx//2-1:nx//2+2,nz//2-1:nz//2+2] = impulse
+
+# Advance simulation for one time step
```
Comment (dpgrote) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: This comment should be updated.

**Review by dpgrote — APPROVED** [review]
This looks great (except one minor comment)! Thanks for the PR!

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 10 above)
Comment (EZoni) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]:
```suggestion
# Initialize inputs and WarpX instance
```

**Comment by EZoni** [comment]
Thank you for your reviews and suggestions, Dave! CI here should pass after #2403 is merged (the test here is using the new features introduced there).

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+        ax.annotate('PML ghost', xy = (nx//2,nz-nzg//2), rotation = 'horizontal', ha = 'center', va = 'center')
+        ax.annotate('PML ghost', xy = (nx//2,nz+nzpml+nzg//2), rotation = 'horizontal', ha = 'center', va = 'center')
+        # Set extent and sliced data
+        extent = np.array([-nxg-nxpml, nx+nxpml+nxg, -nzg-nzpml, nz+nzpml+nzg])
+    else:
+        # Draw ghost regions
+        ax.axvline(x = 0       , linewidth = lw, linestyle = ls)
+        ax.axvline(x = nx      , linewidth = lw, linestyle = ls)
+        ax.axhline(y = 0       , linewidth = lw, linestyle = ls)
+        ax.axhline(y = nz      , linewidth = lw, linestyle = ls)
```
Comment (EZoni) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: White spaces left over:
```suggestion
        ax.axvline(x = 0 , linewidth = lw, linestyle = ls)
        ax.axvline(x = nx, linewidth = lw, linestyle = ls)
        ax.axhline(y = 0 , linewidth = lw, linestyle = ls)
        ax.axhline(y = nz, linewidth = lw, linestyle = ls)
```

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    vmin = -vmax
+    return (vmin, vmax)
+
+# Plot fields data either in valid domain or in PML
+def plot_data(data, pml, title, name):
+    fig, ax = plt.subplots(nrows = 1, ncols = 1, gridspec_kw = dict(wspace = 0.5), figsize = [6,5])
+    cax = make_axes_locatable(ax).append_axes('right', size='5%', pad='5%')
+    lw  = 0.8
+    ls = '--'
+    if (pml):
```
Comment (ax3l) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]:
```suggestion
    if pml:
```

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+sim.write_input_file(file_name = 'inputs_2d')
+
+# Whether to include guard cells in data returned by Python wrappers
+include_ghosts = 1
+
+# Compute min and max of fields data
+def compute_minmax(data):
+    vmax = np.abs(data).max()
+    vmin = -vmax
+    return (vmin, vmax)
```
Comment (ax3l) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]:
```suggestion
    return vmin, vmax
```

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+    impulse_1d = np.array([1./4., 1./2., 1./4.])
+    impulse = np.outer(impulse_1d, impulse_1d)
+    data[nx//2-1:nx//2+2,nz//2-1:nz//2+2] = impulse
+
+# Initialize inputs and WarpX instance
+sim.initialize_inputs()
+sim.initialize_warpx()
+
+# Get fields data using Python wrappers
+import pywarpx.fields as pwxf
```
Comment (ax3l) [inline Examples/Tests/PythonWrappers/PICMI_inputs_2d.py:171]: Let's move this import to the top

**Review by ax3l — COMMENTED** [review]
Looks great to me, tiny updates for the Python logic then it can go in :)

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 12 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Thanks, fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/70ddc6510707c88da1b0eb74bd1f753d50eb3371.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 13 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py]: Thanks, fixed in https://github.com/ECP-WarpX/WarpX/pull/2576/commits/70ddc6510707c88da1b0eb74bd1f753d50eb3371.

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 14 above)
Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py:171]: If I try to move this `import` at the beginning of the file, I actually get a crash (discussed on Slack).

Comment (ax3l) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py:171]: Jeeze, for the record it's:
```
Invalid MIT-MAGIC-COOKIE-1 keyTraceback (most recent call last):
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/_libwarpx.py", line 81, in <module>
    libwarpx = ctypes.CDLL(os.path.join(_get_package_root(), libname))
  File "/usr/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/edoardo/.local/lib/python3.8/site-packages/pywarpx/libwarpx.3d.so: cannot open shared object file: No such file or directory
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "PICMI_inputs_2d.py", line 5, in <module>
    import pywarpx.fields as pwxf
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/fields.py", line 24, in <module>
    from . import _libwarpx
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/_libwarpx.py", line 85, in <module>
    raise Exception(f'"{libname}" was not installed. Installation instructions can be found here https://warpx.readthedocs.io/en/latest/install/users.html') from e
Exception: "libwarpx.3d.so" was not installed. Installation instructions can be found here https://warpx.readthedocs.io/en/latest/install/users.html
Traceback (most recent call last):
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/_libwarpx.py", line 81, in <module>
    libwarpx = ctypes.CDLL(os.path.join(_get_package_root(), libname))
  File "/usr/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/edoardo/.local/lib/python3.8/site-packages/pywarpx/libwarpx.3d.so: cannot open shared object file: No such file or directory
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "PICMI_inputs_2d.py", line 5, in <module>
    import pywarpx.fields as pwxf
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/fields.py", line 24, in <module>
    from . import _libwarpx
  File "/home/edoardo/.local/lib/python3.8/site-packages/pywarpx/_libwarpx.py", line 85, in <module>
    raise Exception(f'"{libname}" was not installed. Installation instructions can be found here https://warpx.readthedocs.io/en/latest/install/users.html') from e
Exception: "libwarpx.3d.so" was not installed. Installation instructions can be found here https://warpx.readthedocs.io/en/latest/install/users.html
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:
  Process name: [[31374,1],1]
  Exit code:    1
--------------------------------------------------------------------------
```

Haven't seen that, but it's not good.

**Review by ax3l — APPROVED** [review]

File: Examples/Tests/PythonWrappers/PICMI_inputs_2d.py (hunk 14 above)
Comment (ax3l) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py:171]: Needs an issue :)

Comment (EZoni) [reply Examples/Tests/PythonWrappers/PICMI_inputs_2d.py:171]: Here it is: #2633.
